tapfreighter/wallet.go

```diff
@@ -511,6 +521,141 @@ func (f *AssetWallet) FundPacket(ctx context.Context,
 	return f.fundPacketWithInputs(ctx, fundDesc, vPkt, selectedCommitments)
 }
 
+// FundBurn funds a virtual transaction for burning the given amount of units of
+// the given asset.
+func (f *AssetWallet) FundBurn(ctx context.Context,
+	fundDesc *tapscript.FundingDescriptor) (*FundedVPacket, error) {
+
+	// We need to find a commitment that has enough assets to satisfy this
+	// send request. We'll map the address to a set of constraints, so we
+	// can use that to do Taproot asset coin selection.
+	constraints := CommitmentConstraints{
+		GroupKey: fundDesc.GroupKey,
+		AssetID:  &fundDesc.ID,
+		MinAmt:   fundDesc.Amount,
+	}
+	selectedCommitments, err := f.cfg.CoinSelector.SelectCoins(
+		ctx, constraints, PreferMaxAmount,
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	// If we return with an error, we want to release the coins we've
+	// selected.
+	success := false
+	defer func() {
+		if !success {
+			outpoints := fn.Map(
+				selectedCommitments,
+				func(c *AnchoredCommitment) wire.OutPoint {
+					return c.AnchorPoint
+				},
+			)
+			err := f.cfg.CoinSelector.ReleaseCoins(
+				ctx, outpoints...,
+			)
+			if err != nil {
+				log.Errorf("Unable to release coins: %v", err)
+			}
+		}
+	}()
+
+	activeAssets := fn.Filter(
+		selectedCommitments, func(c *AnchoredCommitment) bool {
+			return c.Asset.ID() == fundDesc.ID
+		},
+	)
+
+	// Now that we know what inputs we're going to spend, we know that by
+	// definition, we use the first input's info as the burn's PrevID.
+	firstInput := activeAssets[0]
+	firstPrevID := asset.PrevID{
+		OutPoint: firstInput.AnchorPoint,
+		ID:       firstInput.Asset.ID(),
+		ScriptKey: asset.ToSerialized(
+			firstInput.Asset.ScriptKey.PubKey,
+		),
+	}
+	burnKey := asset.NewScriptKey(asset.DeriveBurnKey(firstPrevID))
+	newInternalKey, err := f.cfg.KeyRing.DeriveNextKey(
+		ctx, asset.TaprootAssetsKeyFamily,
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	// We want both the burn output and the change to be in the same anchor
+	// output, that's why we create the packet manually.
+	vPkt := &tappsbt.VPacket{
+		Inputs: []*tappsbt.VInput{{
+			PrevID: asset.PrevID{
+				ID: fundDesc.ID,
+			},
+		}},
+		Outputs: []*tappsbt.VOutput{{
+			Amount:            0,
+			Type:              tappsbt.TypeSplitRoot,
+			AnchorOutputIndex: 0,
+			ScriptKey:         asset.NUMSScriptKey,
```

Comment:
Thought tombstones weren't needed for this case? cc @jharveyb

Reply:
IIUC the tombstone output is added here so that `fundPacketWithInputs` handles change calculation on building a valid change output if needed. The branch later on handles the full-value burn situation, so we should never end up with an actual tombstone. I think the comment could be a bit more verbose in nothing that the final result will not be a tombstone.

Reply:
IIUC the tombstone output is added here so that `fundPacketWithInputs` handles change calculation on building a valid change output if needed. The branch later on handles the full-value burn situation, so we should never end up with an actual tombstone. I think the comment could be a bit more verbose in nothing that the final result will not be a tombstone.

Reply:
Correct, they aren't. This is just to give the wallet an asset output it can attach any passive assets to. We remove it again later if there are no passive assets. Added a more explicit comment to it.